### ISSUE TITLE
Docker Swarm service discovery is taking more time than expected.

### ISSUE THREAD

**Issue by RamanjaneyuluIdavalapati** [moby/moby#37427]
**Description**

I have two servers and both are in docker swarm. I deployed docker stack which has two services in it. Both are running in the same overlay network. One service(**master**) will make calls to other service(**shard**) internally using service name. This call should take about 7 milliseconds on an average. But this is taking **200** to **250** milliseconds some times (out of 100 calls 3 times it is happening).

If I run the services directly on the machines without docker then everything looks fine.

**Steps to reproduce the issue:**

**1. master information:**
`app.py`
```python
import urllib
import socket
import time

from flask import Flask
app = Flask(__name__)

@app.route("/")
def hello():
    shard1_link = "http://shard1"

    t = time.time()
    shard1_content = urllib.urlopen(shard1_link).read()

    html = "<h3>{shard1} {time}</h3>"

    return html.format(shard1=shard1_content, time=time.time()-t)

if __name__ == '__main__':
    app.run(host='0.0.0.0', port=80)
```
`Dockerfile`
```Dockerfile
FROM python:2.7-slim
RUN pip install Flask
ADD . .
EXPOSE 80
CMD ["python", "app.py"]
```
`build_image:` sudo docker build -t master .

**2. shard information:**
`app.py`
```python
import sys
import socket

from flask import Flask

app = Flask(__name__)

@app.route("/")
def hello():
    html = "<h3>{shard}</h3>" \
           "<b>Hostname:</b> {hostname}<br/>"
    return html.format(hostname=socket.gethostname(), shard=sys.argv[1])

if __name__ == "__main__":
    app.run(host='0.0.0.0', port=80)
```
`Dockerfile:`
```Dockerfile
FROM python:2.7-slim
WORKDIR /app
ADD . /app
RUN pip install Flask
EXPOSE 80
CMD ["python", "app.py", "shard1"]
```
`build_image:` sudo docker build -t shard1 .

**3. Deploy stack**
`compose file`
```yml
version: "3.4"
services:
    shard1:
        image: shard1
        deploy:
            placement:
                constraints: [node.hostname == server2]
            replicas: 1
            restart_policy:
                condition: on-failure
        networks:
            - abc
    master:
        image: master
        deploy:
            placement:
                constraints: [node.hostname == server1]
            replicas: 1
            restart_policy:
                condition: on-failure
        ports:
            - 8000:80
        networks:
            - abc
networks:
    abc:
        driver: overlay
```
`deploy:` sudo docker stack deploy -c docker-compose.yml test

**4. Make calls:** for i in {0..100}; do sleep 0.5 && time wget -O - --quiet "http://server1:8000/"; done;

**Describe the results you received:**
Some times making http calls through docker swarm service discovery is taking more time (200 to 250 ms)

**Describe the results you expected:**
All the time the http calls should be consistent (average 7 ms)

**Additional information you deem important (e.g. issue happens only occasionally):**
Issue is happening for 3 times for every 100 calls.

**Output of `docker version`:**

```
Client:
 Version:      18.03.1-ce
 API version:  1.37
 Go version:   go1.9.5
 Git commit:   9ee9f40
 Built:        Thu Apr 26 07:17:20 2018
 OS/Arch:      linux/amd64
 Experimental: false
 Orchestrator: swarm

Server:
 Engine:
  Version:      18.03.1-ce
  API version:  1.37 (minimum version 1.12)
  Go version:   go1.9.5
  Git commit:   9ee9f40
  Built:        Thu Apr 26 07:15:30 2018
  OS/Arch:      linux/amd64
  Experimental: false
```

**Output of `docker info`:**

```
Containers: 8
 Running: 4
 Paused: 0
 Stopped: 4
Images: 16
Server Version: 18.03.1-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay 
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: ypfuijull0sshwsfbuzamumyd
 Is Manager: true
 ClusterID: ffy8v628p1xdfe0g39d615lyf
 Managers: 1
 Nodes: 2
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 10
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Autolock Managers: false
 Root Rotation In Progress: false
 Node Address: X.X.X.X
 Manager Addresses:
  X.X.X.X:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 773c489c9c1b21a6d78b5c538cd395416ec50f88
runc version: 4fc53a81fb7c994640722ac585fa9ca548971871
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.13.0-36-generic
Operating System: Ubuntu 16.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 12
Total Memory: 62.84GiB
Name: hp97.servers.nferx.com
ID: CZMU:RXFU:F5Z6:M6FO:FJBW:K3KS:VICG:LIPK:ROLW:SRDE:HZME:JVMB
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Comment by RamanjaneyuluIdavalapati** [comment]
- Even directly using container IP also causing problem.
- This is not happening when the two services are on the same machine. This is happening when the services are on the different machines.

**Comment by scopev24** [comment]
Hi!

I've seen this problem as well on docker 18.03.01-ce.

What I could find out, that it is not opening the connection that takes that long, but the dns query.
I've seen this on a single node swarm as well as on a three node swarm.

We start a lot of containers every 5 minutes using cron. And every 5 minutes this results in a load spike on the machine. To be more precise, the dockerd process spikes in CPU usage considerably, which seems ok for starting lots of containers at once (about 20-50).

But during that time docker dns resolution slow down. Our system has a resolv.conf with "attempts 2, timeout 2". And seeing dns timeouts means that it took more than 4 seconds to resolve.

The problem went away, when I switched to IP addresses directly - so it's interesting, that you see the problem using IP addresses as well.

I was not able to reproduce this load scenario on a test box - even with exceeded a load of 10 for a single core machine - dockerd name resolution was just fine.

I switched to 18.06.0-rc1 and the problem went away as well - now the single node swarm is again using container dns names for resolution - without any problems.
What is left are now some connection timeouts (on the three node swarm), very sporadic, but I do not know if these are related as they do not occur every 5 minutes.

To sum up: Can you try if you can reproduce with 18.06.0-rc1 (or rc2 which just came out)?
